File: python/app/token/staking.py
# app/token/staking.py

from __future__ import annotations

import time
from dataclasses import dataclass, field
from enum import Enum

from app.config import(
MIN_STAKE,
SLASH_PERCENT,
SLASH_BURN_RATIO,
SLASH_POOL_RATIO,
MAX_VIOLATIONS
)

class StakeStatus(Enum):
    ACTIVE = "active"       
    SLASHED = "slashed"   
    EJECTED = "ejected"      
    WITHDRAWN = "withdrawn"  


class ViolationType(Enum):
    DOUBLE_VOTE = "double_vote"          
    CONFLICTING_TX = "conflicting_tx"    
    REPUTATION_PENALTY = "reputation_penalty" 
    INVALID_STATE = "invalid_state"      


@dataclass
class StakeRecord:
    """Запись о стейке одного узла."""
    address: str
    amount: int                      
    original_amount: int                
    staked_at: float = field(default_factory=time.time)
    status: StakeStatus = StakeStatus.ACTIVE
    violations: list[dict] = field(default_factory=list)
    total_slashed: int = 0
    last_active: float = field(default_factory=time.time)

    def is_active(self) -> bool:
        return self.status == StakeStatus.ACTIVE

    def violation_count(self) -> int:
        return len(self.violations)

    def stake_ratio(self) -> float:
        """Какая доля от начального стейка осталась."""
        if self.original_amount == 0:
            return 0.0
        return self.amount / self.original_amount


@dataclass
class SlashResult:
    slashed_amount: int
    burned: int
    to_pool: int
    ejected: bool
    reason: str


class StakingManager:
    def __init__(self) -> None:
        self.stakes: dict[str, StakeRecord] = {}
        self.slash_pool: int = 0       
        self.total_burned: int = 0   

    def stake(self, address: str, amount: int, token_balances: dict) -> tuple[bool, str]:
        if amount < MIN_STAKE:
            return False, f"minimum stake is {MIN_STAKE} GHOST"

        if address in self.stakes and self.stakes[address].is_active():
            return False, "already staking"

        current_balance = token_balances.get(address, 0)
        if current_balance < amount:
            return False, "insufficient balance"

        token_balances[address] = current_balance - amount

        self.stakes[address] = StakeRecord(
            address=address,
            amount=amount,
            original_amount=amount,
        )
        return True, "staked"

    def slash(
        self,
        address: str,
        violation: ViolationType,
        evidence: str = "",
    ) -> SlashResult | None:
        if address not in self.stakes:
            return None

        record = self.stakes[address]
        if record.status == StakeStatus.EJECTED or record.status == StakeStatus.WITHDRAWN:
            return None

        record.violations.append({
            "type": violation.value,
            "evidence": evidence,
            "timestamp": time.time(),
        })

        slash_amount = int(record.amount * SLASH_PERCENT)
        slash_amount = min(slash_amount, record.amount)

        burned = int(slash_amount * SLASH_BURN_RATIO)
        to_pool = slash_amount - burned

        record.amount -= slash_amount
        record.total_slashed += slash_amount
        record.status = StakeStatus.SLASHED
        self.total_burned += burned
        self.slash_pool += to_pool

        ejected = False

        if record.violation_count() >= MAX_VIOLATIONS:
            remaining = record.amount
            burned_remaining = int(remaining * SLASH_BURN_RATIO)
            pool_remaining = remaining - burned_remaining

            self.total_burned += burned_remaining
            self.slash_pool += pool_remaining
            record.amount = 0
            record.status = StakeStatus.EJECTED
            ejected = True
        else:
            if record.amount >= MIN_STAKE:
                record.status = StakeStatus.ACTIVE

        return SlashResult(
            slashed_amount=slash_amount,
            burned=burned,
            to_pool=to_pool,
            ejected=ejected,
            reason=violation.value,
        )

    def withdraw(
        self,
        address: str,
        token_balances: dict,
    ) -> tuple[bool, str, int]:
        if address not in self.stakes:
            return False, "not staking", 0

        record = self.stakes[address]

        if record.status == StakeStatus.EJECTED:
            return False, "ejected nodes cannot withdraw", 0

        if record.status == StakeStatus.WITHDRAWN:
            return False, "already withdrawn", 0

        amount = record.amount
        token_balances[address] = token_balances.get(address, 0) + amount
        record.amount = 0
        record.status = StakeStatus.WITHDRAWN

        return True, "withdrawn", amount

    def distribute_slash_pool(self, token_balances: dict) -> int:
        if self.slash_pool == 0:
            return 0

        active_nodes = [
            addr for addr, record in self.stakes.items()
            if record.is_active() and record.violation_count() == 0
        ]

        if not active_nodes:
            return 0

        per_node = self.slash_pool // len(active_nodes)
        if per_node == 0:
            return 0

        distributed = 0
        for addr in active_nodes:
            token_balances[addr] = token_balances.get(addr, 0) + per_node
            distributed += per_node

        self.slash_pool -= distributed
        return distributed

    def is_eligible(self, address: str) -> bool:
        if address not in self.stakes:
            return False
        record = self.stakes[address]
        return record.is_active() and record.amount >= MIN_STAKE

    def get_stake_weight(self, address: str) -> float:
        if not self.is_eligible(address):
            return 0.0
        record = self.stakes[address]
        return record.stake_ratio()

    def stats(self) -> dict:
        total_staked = sum(r.amount for r in self.stakes.values())
        active = sum(1 for r in self.stakes.values() if r.is_active())
        ejected = sum(1 for r in self.stakes.values() if r.status == StakeStatus.EJECTED)
        return {
            "total_stakers": len(self.stakes),
            "active_stakers": active,
            "ejected": ejected,
            "total_staked": total_staked,
            "slash_pool": self.slash_pool,
            "total_burned": self.total_burned,
            "min_stake": MIN_STAKE,
        }

File: Rust/ghost_core/token/src/staking.rs
// token/src/staking.rs

use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

pub const MIN_STAKE: u64 = 1_000;
const SLASH_PERCENT: f64 = 0.10;
const SLASH_BURN_RATIO: f64 = 0.50;
const MAX_VIOLATIONS: usize = 3;

pub const MIN_VALIDATOR_STAKE: u64 = MIN_STAKE;

pub const MIN_REWARD_STAKE: u64 = MIN_STAKE / 2;

fn now_secs() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
}

#[derive(Debug, Clone, PartialEq)]
pub enum StakeStatus {
    Active,
    Slashed,
    Ejected,
    Withdrawn,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ViolationType {
    DoubleVote,
    ConflictingTx,
    ReputationPenalty,
    InvalidState,
}

impl ViolationType {
    pub fn as_str(&self) -> &str {
        match self {
            ViolationType::DoubleVote       => "double_vote",
            ViolationType::ConflictingTx    => "conflicting_tx",
            ViolationType::ReputationPenalty => "reputation_penalty",
            ViolationType::InvalidState     => "invalid_state",
        }
    }
}

#[derive(Debug, Clone)]
pub struct StakeRecord {
    pub address: String,
    pub amount: u64,
    pub original_amount: u64,
    pub staked_at: f64,
    pub status: StakeStatus,
    pub violations: Vec<String>,
    pub total_slashed: u64,
}

impl StakeRecord {
    pub fn is_active(&self) -> bool {
        self.status == StakeStatus::Active
    }

    pub fn violation_count(&self) -> usize {
        self.violations.len()
    }

    pub fn stake_ratio(&self) -> f64 {
        if self.original_amount == 0 { return 0.0; }
        self.amount as f64 / self.original_amount as f64
    }

    pub fn is_validator(&self) -> bool {
        self.is_active() && self.amount >= MIN_VALIDATOR_STAKE
    }

    pub fn is_reward_eligible(&self) -> bool {
        self.is_active() && self.amount >= MIN_REWARD_STAKE
    }
}

#[derive(Debug)]
pub struct SlashResult {
    pub slashed_amount: u64,
    pub burned: u64,
    pub to_pool: u64,
    pub ejected: bool,
    pub reason: String,
}

#[derive(Debug, PartialEq)]
pub enum EligibilityStatus {
    Validator,
    RewardOnly,
    RelayOnly,
    Ejected,
}

#[derive(Debug, Default)]
pub struct StakingManager {
    pub stakes: HashMap<String, StakeRecord>,
    pub slash_pool: u64,
    pub total_burned: u64,
}

impl StakingManager {
    pub fn new() -> Self {
        StakingManager::default()
    }

    pub fn stake(
        &mut self,
        address: &str,
        amount: u64,
        balances: &mut HashMap<String, u64>,
    ) -> Result<(), String> {
        if amount < MIN_STAKE {
            return Err(format!("minimum stake is {} GHOST", MIN_STAKE));
        }

        if let Some(r) = self.stakes.get(address) {
            if r.is_active() {
                return Err("already staking".to_string());
            }
        }

        let balance = *balances.get(address).unwrap_or(&0);
        if balance < amount {
            return Err("insufficient balance".to_string());
        }

        *balances.entry(address.to_string()).or_insert(0) -= amount;

        self.stakes.insert(address.to_string(), StakeRecord {
            address: address.to_string(),
            amount,
            original_amount: amount,
            staked_at: now_secs(),
            status: StakeStatus::Active,
            violations: vec![],
            total_slashed: 0,
        });

        Ok(())
    }

    pub fn slash(
        &mut self,
        address: &str,
        violation: ViolationType,
        evidence: &str,
    ) -> Option<SlashResult> {
        let record = self.stakes.get_mut(address)?;

        if matches!(record.status, StakeStatus::Ejected | StakeStatus::Withdrawn) {
            return None;
        }

        record.violations.push(format!("{}:{}", violation.as_str(), evidence));
        record.status = StakeStatus::Slashed;

        let slash_amount = ((record.amount as f64 * SLASH_PERCENT) as u64).min(record.amount);
        let burned = (slash_amount as f64 * SLASH_BURN_RATIO) as u64;
        let to_pool = slash_amount - burned;

        record.amount -= slash_amount;
        record.total_slashed += slash_amount;
        self.total_burned += burned;
        self.slash_pool += to_pool;

        let mut ejected = false;

        if record.violation_count() >= MAX_VIOLATIONS {
            let remaining = record.amount;
            let burned_remaining = (remaining as f64 * SLASH_BURN_RATIO) as u64;
            let pool_remaining = remaining - burned_remaining;
            self.total_burned += burned_remaining;
            self.slash_pool += pool_remaining;
            record.amount = 0;
            record.status = StakeStatus::Ejected;
            ejected = true;
        } else if record.amount >= MIN_STAKE {
            record.status = StakeStatus::Active;
        }

        Some(SlashResult { slashed_amount: slash_amount, burned, to_pool, ejected, reason: violation.as_str().to_string() })
    }

    pub fn withdraw(
        &mut self,
        address: &str,
        balances: &mut HashMap<String, u64>,
    ) -> Result<u64, String> {
        let record = self.stakes.get_mut(address)
            .ok_or("not staking")?;

        if record.status == StakeStatus::Ejected {
            return Err("ejected nodes cannot withdraw".to_string());
        }
        if record.status == StakeStatus::Withdrawn {
            return Err("already withdrawn".to_string());
        }

        let amount = record.amount;
        *balances.entry(address.to_string()).or_insert(0) += amount;
        record.amount = 0;
        record.status = StakeStatus::Withdrawn;
        Ok(amount)
    }

    pub fn eligibility(&self, address: &str) -> EligibilityStatus {
        match self.stakes.get(address) {
            None => EligibilityStatus::RelayOnly,
            Some(r) => match r.status {
                StakeStatus::Ejected => EligibilityStatus::Ejected,
                StakeStatus::Withdrawn => EligibilityStatus::RelayOnly,
                _ if r.amount >= MIN_VALIDATOR_STAKE && r.is_active() => {
                    EligibilityStatus::Validator
                }
                _ if r.amount >= MIN_REWARD_STAKE && r.is_active() => {
                    EligibilityStatus::RewardOnly
                }
                _ => EligibilityStatus::RelayOnly,
            }
        }
    }

    pub fn is_eligible(&self, address: &str) -> bool {
        self.eligibility(address) == EligibilityStatus::Validator
    }

    pub fn is_reward_eligible(&self, address: &str) -> bool {
        matches!(
            self.eligibility(address),
            EligibilityStatus::Validator | EligibilityStatus::RewardOnly
        )
    }

    pub fn get_stake_amount(&self, address: &str) -> f64 {
        self.stakes.get(address)
            .filter(|r| r.is_active())
            .map(|r| r.amount as f64)
            .unwrap_or(0.0)
    }

    pub fn total_stake(&self) -> f64 {
        self.stakes.values()
            .filter(|r| r.is_active())
            .map(|r| r.amount as f64)
            .sum()
    }

    pub fn active_validators(&self) -> HashMap<String, f64> {
        self.stakes.values()
            .filter(|r| r.is_validator())
            .map(|r| (r.address.clone(), r.amount as f64))
            .collect()
    }

    pub fn distribute_slash_pool(
        &mut self,
        balances: &mut HashMap<String, u64>,
    ) -> u64 {
        if self.slash_pool == 0 { return 0; }

        let clean_nodes: Vec<String> = self.stakes.values()
            .filter(|r| r.is_active() && r.violation_count() == 0)
            .map(|r| r.address.clone())
            .collect();

        if clean_nodes.is_empty() { return 0; }

        let per_node = self.slash_pool / clean_nodes.len() as u64;
        if per_node == 0 { return 0; }

        let distributed = per_node * clean_nodes.len() as u64;
        for addr in &clean_nodes {
            *balances.entry(addr.clone()).or_insert(0) += per_node;
        }
        self.slash_pool -= distributed;
        distributed
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_balances(address: &str, amount: u64) -> HashMap<String, u64> {
        let mut m = HashMap::new();
        m.insert(address.to_string(), amount);
        m
    }

    #[test]
    fn test_stake_success() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 5000);
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        assert_eq!(balances["node1"], 5000 - MIN_STAKE);
        assert_eq!(manager.stakes["node1"].amount, MIN_STAKE);
    }

    #[test]
    fn test_stake_below_minimum() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 5000);
        assert!(manager.stake("node1", MIN_STAKE - 1, &mut balances).is_err());
    }

    #[test]
    fn test_stake_insufficient_balance() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 500);
        assert!(manager.stake("node1", MIN_STAKE, &mut balances).is_err());
    }

    #[test]
    fn test_slash_reduces_stake() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 5000);
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        let result = manager.slash("node1", ViolationType::DoubleVote, "").unwrap();
        let expected = (MIN_STAKE as f64 * SLASH_PERCENT) as u64;
        assert_eq!(result.slashed_amount, expected);
    }

    #[test]
    fn test_slash_ejection_after_max_violations() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 10000);
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        let mut result = None;
        for _ in 0..MAX_VIOLATIONS {
            result = manager.slash("node1", ViolationType::DoubleVote, "");
        }
        assert!(result.unwrap().ejected);
        assert_eq!(manager.stakes["node1"].amount, 0);
    }

    #[test]
    fn test_withdraw_returns_stake() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 5000);
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        let amount = manager.withdraw("node1", &mut balances).unwrap();
        assert_eq!(amount, MIN_STAKE);
        assert_eq!(balances["node1"], 5000);
    }

    #[test]
    fn test_withdraw_ejected_fails() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 10000);
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        for _ in 0..MAX_VIOLATIONS {
            manager.slash("node1", ViolationType::DoubleVote, "");
        }
        assert!(manager.withdraw("node1", &mut balances).is_err());
    }

    #[test]
    fn test_distribute_slash_pool() {
        let mut manager = StakingManager::new();
        let mut balances = HashMap::new();
        balances.insert("node1".to_string(), 5000u64);
        balances.insert("node2".to_string(), 5000u64);
        balances.insert("bad".to_string(), 5000u64);
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        manager.stake("node2", MIN_STAKE, &mut balances).unwrap();
        manager.stake("bad", MIN_STAKE, &mut balances).unwrap();
        manager.slash("bad", ViolationType::DoubleVote, "");
        let b1_before = balances["node1"];
        let distributed = manager.distribute_slash_pool(&mut balances);
        assert!(distributed > 0);
        assert!(balances["node1"] > b1_before);
    }

    #[test]
    fn test_is_eligible_requires_min_stake() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 5000);
        assert!(!manager.is_eligible("node1"));
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        assert!(manager.is_eligible("node1"));
    }

    #[test]
    fn test_no_stake_is_relay_only() {
        let manager = StakingManager::new();
        assert_eq!(manager.eligibility("unknown"), EligibilityStatus::RelayOnly);
    }

    #[test]
    fn test_full_stake_is_validator() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 5000);
        manager.stake("node1", MIN_VALIDATOR_STAKE, &mut balances).unwrap();
        assert_eq!(manager.eligibility("node1"), EligibilityStatus::Validator);
    }

    #[test]
    fn test_ejected_node_is_ejected() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 10000);
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        for _ in 0..MAX_VIOLATIONS {
            manager.slash("node1", ViolationType::DoubleVote, "");
        }
        assert_eq!(manager.eligibility("node1"), EligibilityStatus::Ejected);
    }

    #[test]
    fn test_is_reward_eligible_with_validator_stake() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 5000);
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        assert!(manager.is_reward_eligible("node1"));
    }

    #[test]
    fn test_ejected_not_reward_eligible() {
        let mut manager = StakingManager::new();
        let mut balances = make_balances("node1", 10000);
        manager.stake("node1", MIN_STAKE, &mut balances).unwrap();
        for _ in 0..MAX_VIOLATIONS {
            manager.slash("node1", ViolationType::DoubleVote, "");
        }
        assert!(!manager.is_reward_eligible("node1"));
    }

    #[test]
    fn test_total_stake_sums_active() {
        let mut manager = StakingManager::new();
        let mut balances = HashMap::new();
        balances.insert("a".to_string(), 5000u64);
        balances.insert("b".to_string(), 5000u64);
        manager.stake("a", MIN_STAKE, &mut balances).unwrap();
        manager.stake("b", MIN_STAKE, &mut balances).unwrap();
        assert_eq!(manager.total_stake(), (MIN_STAKE * 2) as f64);
    }

    #[test]
    fn test_get_stake_amount_inactive_is_zero() {
        let manager = StakingManager::new();
        assert_eq!(manager.get_stake_amount("nobody"), 0.0);
    }

    #[test]
    fn test_active_validators_excludes_slashed() {
        let mut manager = StakingManager::new();
        let mut balances = HashMap::new();
        balances.insert("good".to_string(), 5000u64);
        balances.insert("bad".to_string(), 10000u64);
        manager.stake("good", MIN_STAKE, &mut balances).unwrap();
        manager.stake("bad", MIN_STAKE, &mut balances).unwrap();
        for _ in 0..MAX_VIOLATIONS {
            manager.slash("bad", ViolationType::DoubleVote, "");
        }
        let validators = manager.active_validators();
        assert!(validators.contains_key("good"));
        assert!(!validators.contains_key("bad"));
    }
}